### PR TITLE
Breaked down login::index() to several methods. Added global variables to templating.

### PR DESCRIPTION
**Pull request by smilesrg**
In this PR I did the following:

* Break login::index() to several methods. This can help to understand what's going on during the login process, also it could be easier to maintain/make changes.
* There is no need to set $data with language strings: language string can be retrieved in the template by calling ```$lng->get('language_string')```. So all code pieces like ```$data['entry_email'] = $this->language->get('entry_email');``` have been removed from controller.
* Also, ```url``` ```load``` and ```request``` services have been added.
* POST variables can be retrieved directly in the template by calling ```$request->post('variable_name', 'default_value');``` or just ```$request->post('variable_name')```
All code pieces like below have been removed
```php
if (isset($this->request->post['email'])) {
    $data['email'] = $this->request->post['email'];
} else {
	$data['email'] = '';
}
 ```

### CODE REVIEW

**Comment by smilesrg** [comment]
And I want to do more - refactor this login controller by using event listener. But this issue really blocks me https://github.com/opencart/opencart/issues/3233 Code parts like restoring customer cart and wishlist, adding activity, overriding admin login and loading default address can be moved to event listeners.

**Comment by smilesrg** [comment]
@danielkerr I'm really waiting for your opinion on my suggestion about event listeners and this PR itself :-)

**Comment by smilesrg** [comment]
@jamesallsup maybe you can tell your opinion? I really want to participate in OpenCart improvement!

File: upload/catalog/controller/account/login.php [hunk 1]
```diff
@@ -51,95 +19,153 @@ public function index() {
...
+		unset($this->session->data['guest']);
 
-		$data['text_new_customer'] = $this->language->get('text_new_customer');
-		$data['text_register'] = $this->language->get('text_register');
-		$data['text_register_account'] = $this->language->get('text_register_account');
-		$data['text_returning_customer'] = $this->language->get('text_returning_customer');
-		$data['text_i_am_returning_customer'] = $this->language->get('text_i_am_returning_customer');
-		$data['text_forgotten'] = $this->language->get('text_forgotten');
+		// Restore customers cart
+		if ($this->customer->getCart()) {
```
Comment (smilesrg) [inline upload/catalog/controller/account/login.php]: This actually can be moved to event

File: upload/catalog/controller/account/login.php [hunk 2]
```diff
@@ -51,95 +19,153 @@ public function index() {
...
+			foreach ($this->customer->getCart() as $key => $value) {
+				$this->cart->add($key, $value);
+			}
+		}
+
+		// Restore customers wish list
+		if ($this->customer->getWishlist()) {
```
Comment (smilesrg) [inline upload/catalog/controller/account/login.php]: And this can be moved to event

File: upload/catalog/controller/account/login.php [hunk 3]
```diff
@@ -169,60 +191,16 @@ public function index() {
...
-
-		if (!$this->error) {
-			if (!$this->customer->login($this->request->post['email'], $this->request->post['password'])) {
-				$this->error['warning'] = $this->language->get('error_login');
-
-				$this->model_account_customer->addLoginAttempt($this->request->post['email']);
-			} else {
-				$this->model_account_customer->deleteLoginAttempts($this->request->post['email']);
-			}
+	private function loadDefaultAddress()
```
Comment (smilesrg) [inline upload/catalog/controller/account/login.php]: This can be moved to event

File: upload/system/engine/loader.php [hunk 4]
```diff
@@ -62,22 +62,28 @@ public function view($template, $data = array()) {
...
+		}
 
-			require($file);
+		if (isset($data['url']) || isset($data['lng']) || isset($data['load'])) {
+			trigger_error('Error: you must not override url, lng, load and request parameters in the $data array!');
+			exit();
+		}
 
-			$output = ob_get_contents();
+		$data['url'] = $this->registry->get('url');
```
Comment (smilesrg) [inline upload/system/engine/loader.php]: This is "global" variables that can be accessed from any template

**Comment by milimar** [comment]
It is very obvious that you are a very valuable member of the OC community, and you CERTAINLY deserve an answer. 
OC needs you. That is my opinion as a member of the OC community.

**Comment by smilesrg** [comment]
@milimar  thank you! Yesterday I played with branches, and this PR was automatically closed. I'll open a new PR instead, but only after this PR is accepted https://github.com/opencart/opencart/pull/3265
Also, I plan to add priority to event dispatcher. And only then I'll start refactoring of login page and make it even better.

**Comment by 4pxs** [comment]
It will not work in template, where loaded other module. It will always be overridden by the variable. For example 'heading_title' in module and at page. Variable will be taken from the last loaded language file. For example, instead of "Heading Checkout" will appear "Heading module Also Purchased"

**Pull request closed without merging**